libmscore/rendermidi.cpp

```diff
@@ -439,49 +531,449 @@ static void collectMeasureEvents(EventMap* events, Measure* m, Staff* staff, int
                               collectNote(events, channel, note, velocity, tickOffset, staffIdx);
                  }
             }
+      }
 
-      //
-      // collect program changes and controller
-      //
-      for (Segment* s = m->first(SegmentType::ChordRest); s; s = s->next(SegmentType::ChordRest)) {
-            // int tick = s->tick();
-            for (Element* e : s->annotations()) {
-                  if (!e->isStaffTextBase() || e->staffIdx() < firstStaffIdx || e->staffIdx() >= nextStaffIdx)
+//---------------------------------------------------------
+//   changeCCBetween
+//---------------------------------------------------------
+
+static void changeCCBetween(EventMap* events, int stick, int etick, int startExpr, int endExpr, int channel, int controller, VeloChangeMethod changeMethod, int tickOffset)
+      {
+      // Prevent zero-division error, but add single event
+      if (startExpr == endExpr || stick == etick) {
+            NPlayEvent cc11event = NPlayEvent(ME_CONTROLLER, channel, controller, abs(startExpr));
+            events->insert(std::pair<int, NPlayEvent>(stick, cc11event));
+            //qDebug() << "added static cc event val: " << startExpr << " at tick " << stick;
+            return;
+            }
+
+      // Ticks to change expression over
+      int exprTicks = etick-stick;
+      int exprDiff = endExpr - startExpr;
+
+      // The lambdas are in the format:
+      //    func(total CC value difference, current tick, total ticks)
+      // NOTE:JT - possible todo, store constants in functions as const and use capture group
+
+      // NOTE:JT - hardcoded, todo change?
+      int tickInc;
+      if (exprTicks >= MScore::division*2)
+            tickInc = 2;
+      else
+            tickInc = 1;
+
+      // See these functions graphically at: https://www.desmos.com/calculator/kk89ficmjk
+      std::function<int(int,int,int)> valueFunction;
+      switch (changeMethod) {
+            case VeloChangeMethod::EXPONENTIAL:
+                  // Due to the nth-root, exponential functions do not flip with negative values, and cause errors,
+                  // so treat it as a piecewise function.
+                  if (exprDiff > 0)
+                        valueFunction = [](int d, int ct, int tt) { return int(pow(pow((d+1), 1.0/double(tt)), double(ct))-1); };
```

Comment:
Coding rules require spaces around operators. In the following cases, it’s hard enough to read as is, I don’t know if it’d help or hinder. I guess ideal would be to make this more readable in general. Multi line functions or something. Not gonna stress too much though.

Reply:
I've changed it so that it's possibly easier to read now. Or it might be harder. Who knows :)
At least the operators have spacing now.